jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/QoSHandler.java

```diff
@@ -0,0 +1,486 @@
+     * <p>A negative or zero value indicates to calculate
+     * a value based on heuristics, drawn from the number
+     * the size of the server thread pool and/or the number
+     * of CPU cores.</p>
+     *
+     * @param maxRequests the max number of concurrent requests
+     */
+    public void setMaxRequestCount(int maxRequests)
+    {
+        if (isStarted())
+            throw new IllegalStateException("Cannot change maxRequests: " + this);
+        this.maxRequests = maxRequests;
+    }
+
+    /**
+     * @return the max duration of time a request may stay suspended
+     */
+    public Duration getMaxSuspend()
+    {
+        return maxSuspend;
+    }
+
+    /**
+     * <p>Sets the max duration of time a request may stay suspended.</p>
+     * <p>Once the duration expires, the request is failed with an HTTP
+     * status of {@code 503 Service Unavailable}.</p>
+     * <p>{@link Duration#ZERO} means that the request may stay suspended forever.</p>
+     *
+     * @param maxSuspend the max duration of time a request may stay suspended
+     */
+    public void setMaxSuspend(Duration maxSuspend)
+    {
+        if (maxSuspend.isNegative())
+            throw new IllegalArgumentException("invalid maxSuspend duration");
+        this.maxSuspend = maxSuspend;
+    }
+
+    @ManagedAttribute("The number of suspended requests")
+    public long getSuspendedRequestCount()
+    {
+        return state.getLo();
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        timeouts = new Timeouts(getServer().getScheduler());
+        addBean(timeouts);
+
+        int maxRequests = getMaxRequestCount();
+        if (maxRequests <= 0)
+        {
+            ThreadPool threadPool = getServer().getThreadPool();
+            if (threadPool instanceof ThreadPool.SizedThreadPool sized)
+                maxRequests = sized.getMaxThreads() / 2;
+            else
+                maxRequests = ProcessorUtils.availableProcessors();
+            setMaxRequestCount(maxRequests);
+        }
+        state.set(maxRequests, 0);
+
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} initialized maxRequests={}", this, maxRequests);
+
+        super.doStart();
+    }
+
+    @Override
+    protected void doStop() throws Exception
+    {
+        super.doStop();
+        removeBean(timeouts);
+        timeouts.destroy();
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("{} handling {}", this, request);
+
+        while (true)
+        {
+            long encoded = state.get();
+            int permits = AtomicBiInteger.getHi(encoded);
+            if (permits > 0)
+            {
+                if (!state.compareAndSet(encoded, permits - 1, 0))
+                    continue;
+                return forward(request, response, callback);
+            }
```

Comment:
@gregw the CAS will _not_ overwrite the suspended value with `0`.
The invariant we have is that only when `permits` goes to 0, then `suspended` can be incremented, and vice-versa.
So your first suggestion is ok, the second (while working) is not because the invariant is not clear anymore: the 2 values cannot be updated independently, you can only update one if the other is exactly zero.